drivers/usb/device/usb_dc_native_posix_adapt.c

```diff
@@ -0,0 +1,462 @@
+	if (read != sizeof(hdr->u)) {
+		SYS_LOG_ERR("recv() failed: %s", strerror(errno));
+		return;
+	}
+
+	/* Read also padding */
+	read = recv(connfd, &setup_padding, sizeof(setup_padding), 0);
+	if (read != sizeof(setup_padding)) {
+		SYS_LOG_ERR("recv() failed: %s", strerror(errno));
+		return;
+	}
+
+	usbip_header_dump((void *)hdr);
+
+	/* TODO: unlink */
+}
+
+static int handle_import(const u8_t *desc, int connfd)
+{
+	struct op_common header = {
+		.version = htons(USBIP_VERSION),
+		.code = htons(OP_REP_IMPORT),
+		.status = 0,
+	};
+	char busid[32];
+
+	SYS_LOG_DBG("attach device");
+
+	if (recv(connfd, busid, 32, 0) != sizeof(busid)) {
+		SYS_LOG_ERR("recv() failed: %s", strerror(errno));
+		return errno;
+	}
+
+	if (send(connfd, &header, sizeof(header), 0) != sizeof(header)) {
+		SYS_LOG_ERR("send() header failed: %s", strerror(errno));
+		return errno;
+	}
+
+	send_device(desc, connfd);
+
+	return 0;
+}
+
+extern struct usb_desc_header __usb_descriptor_start[];
+
+void usbip_start(void)
+{
+	struct sockaddr_in srv;
+	unsigned char attached;
+	int listenfd, connfd;
+	const u8_t *desc;
+	int reuse = 1;
+
+	SYS_LOG_DBG("Starting");
+
+	/*
+	 * Do not use usb_get_device_descriptor();
+	 * to prevent double string fixing
+	 */
+	desc = (const u8_t *)__usb_descriptor_start;
+	if (!desc) {
+		SYS_LOG_ERR("Descriptors are not set");
+		posix_exit(EXIT_FAILURE);
+	}
+
+	listenfd = socket(PF_INET, SOCK_STREAM | SOCK_NONBLOCK, 0);
+	if (listenfd < 0) {
+		SYS_LOG_ERR("socket() failed: %s", strerror(errno));
+		posix_exit(EXIT_FAILURE);
+	}
+
+	if (setsockopt(listenfd, SOL_SOCKET, SO_REUSEADDR,
+		       (const char *)&reuse, sizeof(reuse)) < 0) {
+		SYS_LOG_WRN("setsockopt() failed: %s", strerror(errno));
+	}
+
+	memset(&srv, 0, sizeof(srv));
+	srv.sin_family = AF_INET;
+	srv.sin_addr.s_addr = htonl(INADDR_ANY);
+	srv.sin_port = htons(USBIP_PORT);
+
+	if (bind(listenfd, (struct sockaddr *)&srv, sizeof(srv)) < 0) {
+		SYS_LOG_ERR("bind() failed: %s", strerror(errno));
+		posix_exit(EXIT_FAILURE);
+	}
+
+	if (listen(listenfd, SOMAXCONN) < 0) {
+		SYS_LOG_ERR("listen() failed: %s", strerror(errno));
+		posix_exit(EXIT_FAILURE);
+	}
+
+	while (true) {
+		struct sockaddr_in client_addr;
+		socklen_t client_addr_len = sizeof(client_addr);
+
+		connfd = accept(listenfd, (struct sockaddr *)&client_addr,
+				&client_addr_len);
+		if (connfd < 0) {
+			if (errno == EAGAIN || errno == EWOULDBLOCK) {
+				/* Non-blocking accept */
+				k_sleep(100);
+
+				continue;
+			}
+
+			SYS_LOG_ERR("accept() failed: %s", strerror(errno));
+			posix_exit(EXIT_FAILURE);
+		}
+
+		connfd_global = connfd;
+
+		SYS_LOG_DBG("Connection: %s", inet_ntoa(client_addr.sin_addr));
+
+		/* Set attached 0 */
+		attached = 0;
+
+		while (true) {
+			struct usbip_header cmd;
+			struct usbip_header_common *hdr = &cmd.common;
+			int read;
+
+			if (!attached) {
+				struct op_common req;
+
+				read = recv(connfd, &req, sizeof(req), 0);
+				if (read != sizeof(req)) {
```

Comment:
Note that the `recv()` will also be non blocking, and will need to be wrapped in a loop with a `k_sleep()`

Reply:
Fixed this, there are some `recv` calls which read part of the message, so it is OK to get error when message is broken.